src/backend/distributed/utils/connection_management.c

```diff
@@ -0,0 +1,541 @@
+
+			nextCell = lnext(connectionCell);
+
+			/*
+			 * To avoid code leaking connections we warn if connections are
+			 * still claimed exclusively. We can only do so if the transaction
+			 * committed, as it's normal that code didn't have chance to clean
+			 * up after errors.
+			 */
+			if (isCommit && connection->claimedExclusively)
+			{
+				ereport(WARNING,
+						(errmsg("connection claimed exclusively at transaction commit")));
+			}
+
+			/*
+			 * Only let a connection life longer than a single transaction if
+			 * instructed to do so by the caller. We also skip doing so if
+			 * it's in a state that wouldn't allow us to run queries again.
+			 */
+			if (!connection->sessionLifespan ||
+				PQstatus(connection->conn) != CONNECTION_OK ||
+				PQtransactionStatus(connection->conn) != PQTRANS_IDLE)
+			{
+				PQfinish(connection->conn);
+				connection->conn = NULL;
+
+				entry->connections =
+					list_delete_cell(entry->connections, connectionCell, previousCell);
+
+				pfree(connection);
+			}
+			else
+			{
+				/* reset per-transaction state */
+				connection->activeInTransaction = false;
+
+				UnclaimConnection(connection);
+
+				previousCell = connectionCell;
+			}
+		}
+
+		/*
+		 * NB: We leave the hash entry in place, even if there's no individual
+		 * connections in it anymore. There seems no benefit in deleting it,
+		 * and it'll save a bit of work in the next transaction.
+		 */
+	}
+}
+
+
+/*
+ * GetNodeConnection() establishes a connection to remote node, using default
+ * user and database.
+ *
+ * See StartNodeUserDatabaseConnection for details.
+ */
+MultiConnection *
+GetNodeConnection(uint32 flags, const char *hostname, int32 port)
+{
+	return GetNodeUserDatabaseConnection(flags, hostname, port, NULL, NULL);
+}
+
+
+/*
+ * StartNodeConnection initiate a connection to remote node, using default
+ * user and database.
+ *
+ * See StartNodeUserDatabaseConnection for details.
+ */
+MultiConnection *
+StartNodeConnection(uint32 flags, const char *hostname, int32 port)
+{
+	return StartNodeUserDatabaseConnection(flags, hostname, port, NULL, NULL);
+}
+
+
+/*
+ * GetNodeUserDatabaseConnection establishes connection to remote node.
+ *
+ * See StartNodeUserDatabaseConnection for details.
+ */
+MultiConnection *
+GetNodeUserDatabaseConnection(uint32 flags, const char *hostname, int32 port, const
+							  char *user, const char *database)
+{
+	MultiConnection *connection;
+
+	connection = StartNodeUserDatabaseConnection(flags, hostname, port, user, database);
+
+	FinishConnectionEstablishment(connection);
+
+	return connection;
+}
+
+
+/*
+ * StartNodeUserDatabaseConnection() initiates a connection to a remote node.
+ *
+ * If user or database are NULL, the current session's defaults are used. The
+ * following flags influence connection establishment behaviour:
+ * - NEW_CONNECTION - it is permitted to establish a new connection
+ * - CACHED_CONNECTION - it is permitted to re-use an established connection
```

Comment:
For simplicity, I'd prefer to remove these flags until there is a concrete need for them. As a caller of this API, I don't care that much whether my connection is cached or new. Even in the overall PR, all the callers set both flags right now.

In general, it seems like callers will mainly care a lot about what state the connection is in (e.g. I might want a connection that is not claimed, not part of coordinated transaction, has no custom SETs), so those are probably appropriate flags at some point.